There code can't complie pass though.

```
struct U <T> {
  wtf: Option<Box<U<T>>>
}
fn main() {
 U{ wtf: Some(Box(U{ wtf: None })) };
}
```
```
<anon>:5:19: 5:33 error: mismatched types:
 expected `core::ptr::Unique<_>`,
    found `U<_>`
(expected struct `core::ptr::Unique`,
    found struct `U`) [E0308]
<anon>:5  U{ wtf: Some(Box(U{ wtf: None })) };
                           ^~~~~~~~~~~~~~
<anon>:5:15: 5:34 error: mismatched types:
 expected `Box<U<_>>`,
    found `alloc::boxed::Box<_>`
(expected box,
    found struct `alloc::boxed::Box`) [E0308]
<anon>:5  U{ wtf: Some(Box(U{ wtf: None })) };
                       ^~~~~~~~~~~~~~~~~~~
error: aborting due to 2 previous errors
playpen: application terminated with error code 101
```
Could someone change the bad title? I am not good for english.

That error is correct; but either way, the other error you get today is this, which while not ideal, is possibly the best we can do. I guess we could look for a `new` method with a single parameter? But that seems hacky.

```
error[E0423]: expected function, found struct `Box`
 --> test.rs:5:18
  |
5 |     U{ wtf: Some(Box(U{ wtf: None })) };
  |                  ^^^
  |                  |
  |                  did you mean `Box { /* fields */ }`?
  |                  constructor is not visible here due to private fields

error: aborting due to previous error
```

> I guess we could look for a new method with a single parameter? But that seems hacky.

It is, but it is a common enough pattern and such a big enhancement in end user experience that I feel it'd be worth it to add it, specially when the ctor is not public. Then, the output should be:

```
error[E0423]: expected function, found struct `Box`
 --> test.rs:5:18
  |
5 |     U{ wtf: Some(Box(U{ wtf: None })) };
  |                  ^^^ help: did you mean to use the `new` method?: `Box::new`

error: aborting due to previous error
```

Similar output, but not quite the same cause in #22692.

Current output:

```
error[E0423]: expected function, found struct `Box`
 --> src/main.rs:5:15
  |
5 |  U{ wtf: Some(Box(U{ wtf: None })) };
  |               ^^^ constructor is not visible here due to private fields
help: possible better candidates are found in other modules, you can import them into scope
  |
1 | use syn::Expr::Box;
  |
1 | use syn::Pat::Box;
  |
1 | use syn::token::Box;
  |

error[E0392]: parameter `T` is never used
 --> src/main.rs:1:11
  |
1 | struct U <T> {
  |           ^ unused type parameter
  |
  = help: consider removing `T` or using a marker such as `std::marker::PhantomData`

error: aborting due to 2 previous errors
```

Current output:

```
error[E0423]: cannot initialize a tuple struct which contains private fields
 --> src/main.rs:7:19
  |
7 |         wtf: Some(Box(U {
  |                   ^^^
  |
note: constructor is not visible here due to private fields
 --> /rustc/92c1937a90e5b6f20fa6e87016d6869da363972e/library/alloc/src/boxed.rs:202:14
  |
  = note: private field
  |
  = note: private field
help: consider importing one of these items instead
  |
1 | use syn::Expr::Box;
  |
1 | use syn::Pat::Box;
  |
1 | use syn::token::Box;
  |
```

We still need to look for `Box::new` and suggest it.

--- a/compiler/rustc_resolve/src/late/diagnostics.rs
+++ b/compiler/rustc_resolve/src/late/diagnostics.rs
@@ -1570,7 +1570,26 @@ impl<'a: 'ast, 'ast, 'tcx> LateResolutionVisitor<'a, '_, 'ast, 'tcx> {
                         err.set_primary_message(
                             "cannot initialize a tuple struct which contains private fields",
                         );
-
+                        if !def_id.is_local()
+                            && self
+                                .r
+                                .tcx
+                                .inherent_impls(def_id)
+                                .iter()
+                                .flat_map(|impl_def_id| {
+                                    self.r.tcx.provided_trait_methods(*impl_def_id)
+                                })
+                                .any(|assoc| !assoc.fn_has_self_parameter && assoc.name == sym::new)
+                        {
+                            // FIXME: look for associated functions with Self return type,
+                            // instead of relying only on the name and lack of self receiver.
+                            err.span_suggestion_verbose(
+                                span.shrink_to_hi(),
+                                "you might have meant to use the `new` associated function",
+                                "::new".to_string(),
+                                Applicability::MaybeIncorrect,
+                            );
+                        }
                         // Use spans of the tuple struct definition.
                         self.r.field_def_ids(def_id).map(|field_ids| {
                             field_ids

--- a/tests/ui/privacy/suggest-box-new.fixed
+++ b/tests/ui/privacy/suggest-box-new.fixed
@@ -1,0 +1,15 @@
+// run-rustfix
+#![allow(dead_code)]
+struct U <T> {
+    wtf: Option<Box<U<T>>>,
+    x: T,
+}
+fn main() {
+    U {
+        wtf: Some(Box::new(U { //~ ERROR cannot initialize a tuple struct which contains private fields
+            wtf: None,
+            x: (),
+        })),
+        x: ()
+    };
+}

--- a/tests/ui/privacy/suggest-box-new.rs
+++ b/tests/ui/privacy/suggest-box-new.rs
@@ -1,0 +1,15 @@
+// run-rustfix
+#![allow(dead_code)]
+struct U <T> {
+    wtf: Option<Box<U<T>>>,
+    x: T,
+}
+fn main() {
+    U {
+        wtf: Some(Box(U { //~ ERROR cannot initialize a tuple struct which contains private fields
+            wtf: None,
+            x: (),
+        })),
+        x: ()
+    };
+}

--- a/tests/ui/privacy/suggest-box-new.stderr
+++ b/tests/ui/privacy/suggest-box-new.stderr
@@ -1,0 +1,20 @@
+error[E0423]: cannot initialize a tuple struct which contains private fields
+  --> $DIR/suggest-box-new.rs:9:19
+   |
+LL |         wtf: Some(Box(U {
+   |                   ^^^
+   |
+note: constructor is not visible here due to private fields
+  --> $SRC_DIR/alloc/src/boxed.rs:LL:COL
+   |
+   = note: private field
+   |
+   = note: private field
+help: you might have meant to use the `new` associated function
+   |
+LL |         wtf: Some(Box::new(U {
+   |                      +++++
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0423`.

